src/commands/dockersupport/askDockerStep.ts

```diff
@@ -0,0 +1,24 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { QuickPickOptions } from "vscode";
+import { IActionContext, IAzureQuickPickItem } from "vscode-azureextensionui";
+import { localize } from "../../localize";
+
+/**
+ * Provides top bar prompt UI asking the user if they'd like to use Docker with their cloned local project
+ * @param context - Provides basic actions for functions
+ * @returns - Yes/No response to using Docker with cloned local project
+ */
+export async function prompt(context: IActionContext): Promise<string> {
```

Comment:
OK, I like this flow. My only concern is that users who don't have docker will never know that that is an option. If they don't have docker, should we still display a prompt saying that Docker is recommended and point them towards docker documentation?

Reply:
One more concern is that if users have docker and remote-containers extension and choose to open their project normally, when the new window loads, it will detect the dev container file and again prompt them if they want to "Reopen in Container" so they get asked that twice if they say no

Reply:
Hey @ejizba, just wanted to elaborate a bit on what Valentina was saying.  Basically we know that many users don't have Docker, so we'd like to use this as a way to sort of advertise this option since we think it will help to get rid of some of the version dependency issues that some customers face when setting up their system.  I'm okay not prompting if you think there's another way to achieve a way to "educate" users on that option so long as that other way is smart enough to only show it if it actually meets the criteria for their app (ie Linux and node or python).

Reply:
> should we still display a prompt saying that Docker is recommended and point them towards docker documentation?

Docker is purely optional, though. In fact, the recommended default experience for our extension is _not_ to use Docker.

![128755874-8e9b01e2-02b0-44d0-94c0-716503bea7b8](https://user-images.githubusercontent.com/11282622/129115657-8f1c48cc-a202-41b2-af42-dbaf60905dd3.png)

I think this prompt is the main roadblock at the moment. In many ways, using a dev container _complicates_ the dev experience rather than simplifying it. If users are writing in JavaScript, they are already familiar with Node.js and shouldn't have a problem installing that on their machine (and many will already have it). The only other thing they have to install is the func cli, but we help them with that in our extension and we delay installing it until absolutely necessary because a lot of our functionality doesn't need it. On the flipside, new users probably do _not_ have Docker installed and many are not familiar with containers, so we risk losing them if we push them down that path. If you want to educate users on Docker, I think the prompt needs to make clear that Docker is an optional part of downloading a project and skipping Docker should be the default.

One alternative is to have two entry points like this:
<img width="434" alt="Screen Shot 2021-08-11 at 4 17 26 PM" src="https://user-images.githubusercontent.com/11282622/129115709-3af6f485-cf3e-4e0e-82ee-21175853bc0e.png">

That way users have a "clean" way to download their project without Docker (our recommended path), but it also adds a pretty visible entry point for people who are curious about dev containers. Just one idea I had 🤷‍♂️

Reply:
Thanks @ejizba for the feedback.  I think one problem we're trying to solve is the case where uses are running a different version of node/python on their local box vs what their app is running.  Trying to install or run multiple versions can complicate things for less savvy developers.

That said, I can see the hesitation to trying to using Docker as the default since this is the first time we're trying to utilize it in this way.  For now, I like your suggestion of changing the default option to not using Docker.  @stefanushinardi can you help us with messaging here please?

Also I probably should have checked with you first, but is there a way for us to capture data on which options users choose?  I think it would be good for us to measure how many people go down each path and how successful they are with either path.

Reply:
We try to avoid "concept cliffs" in our wizards - aka prompts where users have no idea what the words even mean and they have to leave VS Code to go research. As one example, we don't prompt about resource groups when creating a function app because it's an Azure-specific concept that many new users are not familiar with. I'm really excited about the "Download Project" feature by itself, but I think the current Docker prompt (regardless of the default) is too big of a concept cliff for a feature geared towards newer users. If you want to keep the prompt, I would feel much more comfortable if we at least check if Docker is installed first, because then it's much more likely they've at least heard of Docker/containers.

> I think one problem we're trying to solve is the case where uses are running a different version of node/python on their local box vs what their app is running. Trying to install or run multiple versions can complicate things for less savvy developers.

We have not seen feedback or evidence that this is a major problem for our users. One of my guesses is that newer users are not authoring complex enough projects where the node/python version matters that much anyways.

> Also I probably should have checked with you first, but is there a way for us to capture data on which options users choose?

We automatically track a lot of stuff in telemetry, but yes you can add custom info on `context.telemetry`

Reply:
Okay, I think that's fine for now.  At some point we do want to have the Docker option point to an actual installer which would pull in the dependency for you.  If/when we get to that point, would you be okay adding that back in even if they don't have Docker?

In any case, for now @valentinavallalta had a suggestion that if we didn't show the prompt, then we could at least show a toast notification message on the bottom right which says something like "Want to run your app in an isolated container locally?  Click here to learn more".  Would you at least be okay with that?  I do think that we need some way of advertising this feature.  Scott is pushing hard for us to figure out how to use Docker containers with Codespaces, but we don't have the same level of discoverability issues there so I think this is a good opportunity for us to gain traction here.

Reply:
yes, this is the outline of that idea, what do you think @ejizba ?
![image](https://user-images.githubusercontent.com/55974511/129280537-69f48ddb-84de-4e97-918a-25a7fb3113a4.png)

Reply:
Hey guys I think we've been pretty clear for a long time now (at least since last winter) that we do not want dev containers to be a featured experience for new users. I'm already compromising by allowing a prompt in the case that Docker _is_ installed, so the answer to both of @ehamai's questions is no. My compromise is intended as a low-risk way to get more telemetry/feedback on an unproven and contentious feature. If the results are positive then yes I would be willing to change my mind on both questions. Of course if the results are negative (as I expect), then we would want to remove the dev container prompt completely.

Reply:
Hey @ejizba , sorry for the delay.  We needed some time to discuss as a team.  I think what you're saying makes sense for now.  So basically we'll only advertise the Docker feature if they actually have Docker installed.  We'll collect feedback and reevaluate our approach if the user data shows that Docker is a good approach to use overall.  Sound good?

Reply:
Yeah I think that's fine